File: leetcode/two-sum.ts
function twoSum(nums: number[], target: number): number[] {
  const seen: Record<number, number> = {};

  for (let i = 0; i < nums.length; i++) {
    const diff = target - nums[i];

    if (seen[diff] !== undefined) {
      return [seen[diff], i];
    }

    seen[nums[i]] = i;
  }

  return [];
};

const result = twoSum([2, 7, 11, 15], 9);
console.log(result);

File: go/kata/algorithms/two_sum.go
package algorithms

func TwoSum(nums []int, target int) [2]int {
	memo := make(map[int]int)

	for i, num := range nums {
		diff := target - num

		if seenIndex, ok := memo[diff]; ok {
			return [2]int{seenIndex, i}
		}

		memo[num] = i
	}

	return [2]int{-1, -1}
}
